### PR TITLE
fix!: Standardize the shape of `Explanation.base_values` of TreeExplainers

### LINKED ISSUE #2140: Waterfall plot .base_values error
Whilst trying to plot a waterfall ploit with my trained model i get the following exeption error:

````
shap.waterfall_plot(shap_values)

Exception: waterfall_plot requires a scalar base_values of the model output as the first parameter, but you have passed 
an array as the first parameter! Try shap.waterfall_plot(explainer.base_values[0], values[0], X[0]) or for multi-output 
models try shap.waterfall_plot(explainer.base_values[0], values[0][0], X[0]).

````

Then when i try to plot using the recommended approach i get the following error:

````
shap.waterfall_plot(shap_values.base_values[0], shap_values[0],X_validate_dense[0] )

AttributeError: 'numpy.ndarray' object has no attribute 'base_values'

````

It seems that the base values are extracted within the function waterfall_plot() even though the recomendation states you should enter the base values as parameter. I presumed it would work seamlessly since the Explanation output looks just like any other example I have found on the internet:

image

 Is it possible to solve this error or does a workaround exist?

### LINKED ISSUE #2255: Waterfall plot (from website example) breaks with sklearn.ensemble.RandomForestRegressor
Drawing a waterfall plot - using the [Tree Ensemble example](https://github.com/slundberg/shap#tree-ensemble-example-xgboostlightgbmcatboostscikit-learnpyspark-models) - fails when using the RandomForestRegressor:

```python
# Reproducible error, code taken from: https://github.com/slundberg/shap with modification in line 6

import sklearn
import xgboost
import shap

# train a Random Forest model
X, y = shap.datasets.boston()
# model = xgboost.XGBRegressor().fit(X, y) # ORIGINAL EXAMPLE
model = sklearn.ensemble.RandomForestRegressor().fit(X, y) # RANDOM FOREST REGRESSOR

# explain the model's predictions using SHAP
# (same syntax works for LightGBM, CatBoost, scikit-learn, transformers, Spark, etc.)
explainer = shap.Explainer(model)
shap_values = explainer(X)

# visualize the first prediction's explanation
shap.plots.waterfall(shap_values[0])
```

This produces the error: 
```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
/tmp/ipykernel_37/3127424203.py in <module>
     17 #shap_values
     18 # visualize the first prediction's explanation
---> 19 shap.plots.waterfall(shap_values[0])

/opt/conda/lib/python3.7/site-packages/shap/plots/_waterfall.py in waterfall(shap_values, max_display, show)
     54     # make sure we only have a single output to explain
     55     if (type(base_values) == np.ndarray and len(base_values) > 0) or type(base_values) == list:
---> 56         raise Exception("waterfall_plot requires a scalar base_values of the model output as the first " \
     57                         "parameter, but you have passed an array as the first parameter! " \
     58                         "Try shap.waterfall_plot(explainer.base_values[0], values[0], X[0]) or " \

Exception: waterfall_plot requires a scalar base_values of the model output as the first parameter, but you have passed an array as the first parameter! Try shap.waterfall_plot(explainer.base_values[0], values[0], X[0]) or for multi-output models try shap.waterfall_plot(explainer.base_values[0], values[0][0], X[0]).
```

**Further explanation:** 
The problem might be that for the Random Forest, shap_values.base_values[0] is a numpy array (of size 1), while Shap expects a number only (which it gets for XGBoost). Look at the last two lines in each case to see the difference.

*XGBoost (from the working example)*:
```python
model = xgboost.XGBRegressor().fit(X, y) # ORIGINAL EXAMPLE
explainer = shap.Explainer(model)
shap_values = explainer(X)
​
print(shap_values[0])
print(shap_values[0].base_values)
print(type(shap_values.base_values[0]))
```
Output:
```
.values =
array([-4.2850167e-01, -6.6636719e-02,  7.7860229e-02, -1.5295845e-03,
       -7.2922713e-01, -2.1700280e+00,  1.9213372e-01, -4.1425934e-01,
       -4.9156108e-01, -4.7296646e-01,  2.5669456e-01, -5.3907130e-02,
        5.7883248e+00], dtype=float32)

.base_values =
22.532942

.data =
array([6.320e-03, 1.800e+01, 2.310e+00, 0.000e+00, 5.380e-01, 6.575e+00,
       6.520e+01, 4.090e+00, 1.000e+00, 2.960e+02, 1.530e+01, 3.969e+02,
       4.980e+00])
22.532942
<class 'numpy.float32'>
```

*Random Forest Regressor (from the failing example)*:
```python
model = sklearn.ensemble.RandomForestRegressor().fit(X, y)
explainer = shap.Explainer(model)
shap_values = explainer(X)
​
print(shap_values[0])
print(shap_values[0].base_values)
print(type(shap_values.base_values[0]))

```
```
.values =
array([ 0.10307071, -0.0303413 ,  0.09707358, -0.00572019, -0.16674241,
       -1.53239688, -0.06641675, -0.20920796, -0.19404617, -0.17280302,
        0.3969092 ,  0.02818953,  4.53953443])

.base_values =
array([22.47789723])

.data =
array([6.320e-03, 1.800e+01, 2.310e+00, 0.000e+00, 5.380e-01, 6.575e+00,
       6.520e+01, 4.090e+00, 1.000e+00, 2.960e+02, 1.530e+01, 3.969e+02,
       4.980e+00])
[22.47789723]
<class 'numpy.ndarray'>
```

I've tried things like
```python
shap_values.base_values = shap_values.base_values.ravel()
```
before running shap.plots.waterfall(shap_values[0]), but I think I'm breaking the object shap_values with that. I've tried the advice from the error message, but don't fully understand it. 

I'm using Kaggle, which has 
xgboost==1.5.0
shap==0.40.0
scikit-learn==0.23.2
pandas==1.3.3
numpy==1.19.5

My apologies in case this has been answered already (or I just messed up). I've searched the forum but didn't find the same issue open or resolved.

### LINKED ISSUE #2855: IndexError: index 0 is out of bounds for axis 0 with size 0
I encountered the following indexing error using SHAP to draw a dependency plot.
This is my code:
`shap.plots.scatter(shap_values[:,"DEM"], color=shap_values)`

I get this error:
`
Traceback (most recent call last):

  File "C:\Users\A1827\AppData\Local\Temp\ipykernel_27060\3083767016.py", line 1, in <cell line: 1>
    shap.plots.scatter(shap_values[:,"DEM"], color=shap_values)

  File "D:\Application\miniconda\lib\site-packages\shap\_explanation.py", line 350, in __getitem__
    t = np.argwhere(np.array(self.output_names) == t)[0][0]

IndexError: index 0 is out of bounds for axis 0 with size 0
`

Maybe it has to do with the matplotlib version? #326
My matplotlib was 3.5.3
My SHAP was 0.41.0

### PR DESCRIPTION
**Pull request by thatlittleboy**
## Overview

Description of the changes proposed in this pull request:

* (!) Standardize the shape of `Explanation.base_values` of `TreeExplainers`, to make it more predictable. In short:
    * We will now no longer get `base_values` arrays of shape `(N,1)`, it will be of shape `(N,)` instead. Otherwise `(N,2)` etc. for multioutput models.
    * Read the background for this change below.
    * An **API-breaking** change.
* Add tests for all the major TreeExplainer types (lightgbm, xgboost, sklearn, catboost, ...) and also tried to cover regression / binary classification / multiclass classification as far as possible.
    * Tests are now **explicitly** testing for the shapes on the `.base_values` and `.values` arrays. 
    * <s>`XGBRegressor` is notably missing from the coverage here, I will add a test for this eventually when I do a overhaul of the xgboost tests. (created an issue #3123 to track this)</s>

The code change is fairly simple, but the implications are potentially big.  
To reviewers: For API breaking changes like this, I would advocate for having at least 2 members approving the PR before the merge, if possible.

---

## Background

We have received a slew of bug reports arising from the inconsistencies of the shapes of `Explanation.base_values`, which causes `IndexError` problems when users input these `Explanation` objects into our plotting functions.

See issues #2255, #2855 , #2140, #1801 for more background and in-depth diagnosis. In short, we have some tree explainers that give a `Explanation.base_values.shape` of `(N,1)` while others give `(N,)`. Which results in users having to write hacks like [this](https://github.com/shap/shap/issues/1801#issuecomment-811091856) or [this](https://github.com/shap/shap/issues/2255#issuecomment-962260490) to get around the inconsistency.

### What is the inconsistency, *exactly*?

Consider the following:

```python
import xgboost
import sklearn
import shap

X, y = shap.datasets.adult(n_points=500)

# === xgboost ===
clf = xgboost.XGBRegressor().fit(X, y)
explainer = shap.Explainer(clf)
explanation = explainer(X)
print(f"XGboost {explanation.base_values.shape=}")  # (500,)
shap.plots.waterfall(explanation[0], show=False)  # no Error!


# === sklearn ===
clf = sklearn.ensemble.RandomForestRegressor().fit(X, y)
explainer = shap.Explainer(clf)
explanation = explainer(X)
print(f"Sklearn {explanation.base_values.shape=}")  # (500,1)
shap.plots.waterfall(explanation[0], show=False)  # throws Error!
```

The code is exactly the same, only a different regressor was used. Yet one throws an error, and the other doesn't.  
See below for the list of affected models (those that used to return a basevalues shape `(N,1)` before this PR).

### What are the alternatives instead of the decisions made in this PR?

1. We can make all of our plotting functions (e.g. `shap.plots.bar`, `shap.plots.waterfall`, ...) all handle this shape inconsistency internally. But that's just a workaround which is prone to failing in the future. See an example PR here: PR #2667.
2. **Note** that only the output of `Tree.__call__()` is standardized here, not `Tree.expected_value`. Which would itself also be a breaking change itself.
    * We can definitely consider standardizing `Tree.expected_value` in the future, but this decision shouldn't make-or-break this PR.
    * The reason I didn't standardize `expected_value` of the TreeExplainer's yet is just that the changes there is potentially more complicated. We have to look into every Tree implementation and ensure the output shapes are consistent. This PR alone is sufficient to close off most of the problems that people are facing with `base_values` shape, and we can take this opportunity to educate users to use `Explanation` objects as *the standard* API instead of using the raw expected value on the Explainer.
    * There is a *related* problem where the various Tree's give inconsistent shapes of the SHAP values array. The biggest offender is the binary classification case with `XGBClassifier` and `LGBMClassifier`. The former returns a `(N,k)` array, while the latter returns a list of `(N,k)` arrays. For other examples, scan through the tests in `test_tree.py` and compare the `explanation.values.shape` across the various scenarios. I think if we ever standardize `Tree.expected_value`, it should be executed/standardized together with the `Tree.shap_values()` output.

### Which models are *actually* affected?

First of all, it's important to note that none of the other Explainers are affected, only `TreeExplainer`.  

These are the models & tests that changed from failing -> passing when we made this breaking change:

- Model: `sklearn.ensemble.GradientBoostingClassifier`
    - `tests/explainers/test_tree.py::TestExplainerSklearn::test_sum_match_gradient_boosting_classifier`
- Model: `sklearn.ensemble.GradientBoostingRegressor`
    - `tests/explainers/test_tree.py::TestExplainerSklearn::test_sum_match_gradient_boosting_regressor`
- Model: `sklearn.ensemble.HistGradientBoostingRegressor`
    - `tests/explainers/test_tree.py::TestExplainerSklearn::test_HistGradientBoostingRegressor`
-  Model: `sklearn.ensemble.IsolationForest`
    - `tests/explainers/test_tree.py::test_isolation_forest`
- Model: `pyod.models.iforest`
    - `tests/explainers/test_tree.py::test_pyod_isolation_forest`
- Model: `ngboost.NGBRegressor`
    - `tests/explainers/test_tree.py::test_ngboost`
- Model: `gpboost.GPModel`
    - `tests/explainers/test_tree.py::test_gpboost`

All of the above models & associated TreeExplainers had their `Explanation.base_values.shape` change from `(N,1)` to `(N,)`.  
As far as I could tell, the other supported Tree models are not affected.

### What is the final benefit / result?

All TreeExplainer's models should be able to run the example at the beginning of the post now, without having to fumble around with the array shapes. 🎉

## Checklist

- [X] Closes #2255, closes #2855 , closes #2140. Related to #1420, #1801. Also, we can close #2667 after this PR is merged.
- [X] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [X] Unit tests added (if fixing a bug or adding a new feature)
- [ ] Added entry to `CHANGELOG.md` (if changes will affect users)

### CODE REVIEW

File: tests/explainers/test_tree.py
Comment (thatlittleboy) [inline tests/explainers/test_tree.py]: The general crux of all of the changes in this `test_tree.py` file is to utilize the `explainer.__call__(X)` instead of `explainer.shap_values(X)`, and explicitly asserting the shape of the base values and SHAP values arrays.

**Comment by connortann** [comment]
Thanks for all your work on this. From what I can see, it looks like a sensible change. Here are some initial comments..

1) As a nitpick, the diff is reasonably large. Ideally, it would be nice to cherry-pick out the non-breaking changes (e.g. style changes and other cleanups) into a separate PR, leaving this one with just the functional changes.

2) I think it would be helpful to explicitly document the class attributes and their shapes, rather than leaving it implicit. Do you think the class-level docstring of `TreeExplainer` is the right place for this?

**Comment by thatlittleboy** [comment]
> As a nitpick, the diff is reasonably large. Ideally, it would be nice to cherry-pick out the non-breaking changes (e.g. style changes and other cleanups) into a separate PR, leaving this one with just the functional changes.

Can you help to point out what changes I should separate out? The way I see it, the breaking change is the changes in `_tree.py`, and the rest are unit test changes to ensure the breaking changes are implemented correctly. (There are *some* rearranging of logic in the tests, but I don't think separating them out would make the diffs much smaller).


> I think it would be helpful to explicitly document the class attributes and their shapes, rather than leaving it implicit. Do you think the class-level docstring of TreeExplainer is the right place for this?

You mean the `.base_values` property on the Explanation class that is being returned by `TreeExplainer.__call__`, right? I think it can go in the docstring of `__call__` , let me see what I can do about that this weekend.
But any further documentation, I think is out of scope (re: your first nitpick).

**Comment by connortann** [comment]
> Can you help to point out what changes I should separate out?

Sure, although I wouldn't want to over-labour the point. Some examples in `test_tree.py` that look like refactoring changes:

- L147
- L151
- L158
- L222
- L352
- L258
- L384
- L427
- L1203

and so on. Only a minor nitpick as I said, I'd be happy to approve as-is :)

> But any further documentation, I think is out of scope (re: your first nitpick)

Touché 😂 

Overall your changes look good to me, and now I've had some time to digest and re-read the diff I'll approve. I agree with your earlier comment, it would be good to get a review from another maintainer for breaking changes.

**Review by connortann — APPROVED** [review]
I support this change. Thank you for your work on this!

**Comment by dsgibbons** [comment]
Hi @thatlittleboy I can take a second look at this. Looks like there is a merge conflict that needs resolving first though?

**Comment by thatlittleboy** [comment]
> Hi @thatlittleboy I can take a second look at this. Looks like there is a merge conflict that needs resolving first though?

Yep resolved the conflict. Thank you!

**Comment by dsgibbons** [comment]
Looks good. You may have mentioned this already, but can you clarify why `(N,)` was selected over `(N,1)`?

**Comment by thatlittleboy** [comment]
> Looks good. You may have mentioned this already, but can you clarify why `(N,)` was selected over `(N,1)`?

@dsgibbons  It's a good question, I think that it's just an API choice. I opted for `(N,)` instead of `(N,1)` because I think it's perhaps just a little more ergonomic to use, e.g. for the Sklearn regression example, it makes more sense to me for the user to pass in `explanation[0]` rather than `explanation[0,0]`. It's a little awkward to explain why the second `0` has to be there, since there isn't a concept of `n_classes` in regression.

**Comment by thatlittleboy** [comment]
Had to re-push the changes and resolve the conflict due to the recently refactors to xgboost tests in PR #3192.

I've also added in the explanation property tests for XGBRegressor's following #3192. (`TestExplainerXGBoost::test_xgboost_regression`).

@dsgibbons Any further thoughts on this PR?

**Comment by dsgibbons** [comment]
Looks good to me, sorry for the delay on this one.

**Review by dsgibbons — APPROVED** [review]

**Pull request merged**
